### PR TITLE
Smoothn5

### PR DESCRIPTION
**Pull request by timspainUCL**
Merge the smoothn commits

### CODE REVIEW

File: kaska/smoothn.py [hunk 1]
```diff
@@ -1,537 +1,617 @@
...
+
+#  GCV score
+# ---
+# function GCVscore = gcv(p)
+def gcv(p, lambda_v, aow, dct_y, is_finite, w_tot, y, nof, noe, smooth_order):
+    """
+    Search the smoothing parameter s that minimizes the GCV score
+    """
+    s = 10 ** p
+    gamma = 1.0 / (1 + (s * np.abs(lambda_v)) ** smooth_order)
```
Comment (ilectra) [inline kaska/smoothn.py]: gamma is calculated in one more place than here - maybe move to its own function/lambda?

File: kaska/smoothn.py [hunk 2]
```diff
@@ -1,537 +1,617 @@
...
+
+
+#  Robust weights
+# function W = robust_weights(r,I,h,wstr)
+def robust_weights(r, i, h, wstr):
+    """
+    Recalculate the weights for robust smoothing.
+    """
+    mad = np.median(np.abs(r[i] - np.median(r[i])))  # median absolute deviation
+    u = np.abs(r / (1.4826 * mad) / np.sqrt(1 - h))  # studentized residuals
```
Comment (ilectra) [inline kaska/smoothn.py]: Maybe provide some references for the hard-coded numbers? (@jgomezdans ?)

File: kaska/smoothn.py [hunk 3]
```diff
@@ -1,537 +1,617 @@
...
-    z[d] = 0.
-    z = dctND(z,f=idct)
+    z[d] = 0.0
+    z = dctND(z, f=idct)
     return z
-    #-- coarse fast smoothing using one-tenth of the DCT coefficients
-    #siz = z.shape;
-    #z = dct(z,norm='ortho',type=2);
-    #for k in np.arange(len(z.shape)):
+    # -- coarse fast smoothing using one-tenth of the DCT coefficients
```
Comment (ilectra) [inline kaska/smoothn.py]: Remove commented code.

File: kaska/smoothn.py [hunk 4]
```diff
@@ -1,537 +1,617 @@
...
+                    ss = np.arange(nS0) * (1.0 / (nS0 - 1.0)) * (
+                        np.log10(s_max_bnd) - np.log10(s_min_bnd)
+                        ) + np.log10(s_min_bnd)
+                    g = np.zeros_like(ss)
+                    for i, p in enumerate(ss):
+                        g[i] = gcv(p, lambda_, aow, dct_y, is_finite, w_tot,
+                                   y, nof, noe, smoothOrder)
+                        # print 10**p,g[i]
+                    xpost = [ss[g == g.min()]]
+                    # print '==============='
```
[the remark this answers is not in the document]
Comment (timspainUCL) [reply kaska/smoothn.py]: Done

File: kaska/smoothn.py (hunk 3 above)
Comment (timspainUCL) [reply kaska/smoothn.py]: Done

File: kaska/smoothn.py (hunk 1 above)
Comment (timspainUCL) [reply kaska/smoothn.py]: Done

File: kaska/smoothn.py [hunk 5]
```diff
@@ -1,537 +1,617 @@
...
+                 norm="ortho", type=2, axis=2)
+    if nd == 4:
+        return f(f(f(f(data, norm="ortho", type=2, axis=0),
+                     norm="ortho", type=2, axis=1),
+                   norm="ortho", type=2, axis=2),
+                 norm="ortho", type=2, axis=3)
+
+
+"""
 def test1():
```
[the remark this answers is not in the document]
Comment (timspainUCL) [reply kaska/smoothn.py]: They are gone

File: kaska/smoothn.py [hunk 6]
```diff
@@ -1,537 +1,617 @@
...
+        )
+
+    return z, s, exitflag, w_tot
+
+
+def warning(s1, s2):
+    """
+    Combine warnings.
+    """
+    print(s1)
```
[the remark this answers is not in the document]
Comment (timspainUCL) [reply kaska/smoothn.py]: No more printing

File: kaska/smoothn.py [hunk 7]
```diff
@@ -1,537 +1,617 @@
...
+EXIT_SUCCESS = 0
+EXIT_LIB_NOT_FOUND = -1
+
+W_TOT_DEFAULT = 0
+
+
+def smoothn(y, nS0=10, axis=None, smoothOrder=2.0, sd=None, verbose=False,
+            s0=None, z0=None, isrobust=False, w=None, s=None, max_iter=100,
+            tol_z=1e-3, weightstr='bisquare'):
+    """
```
[the remark this answers is not in the document]
Comment (timspainUCL) [reply kaska/smoothn.py]: Followed the example

**Comment by timspainUCL** [comment]
@ilectra I think i've resolved all the comments, apart from the mystery weightings numbers.

**Comment by ilectra** [comment]
The rest of the functions (in addition to `smoothn`) also need their proper docstrings. After that, I'm fine to merge.

**Comment by timspainUCL** [comment]
Awwww.
This is why people write 500 line functions, you know.

File: kaska/smoothn.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jgomezdans) [reply kaska/smoothn.py]: Yes, that was the point of this task: to simplify the function so as to optimise it for two use cases (smoothing a 1D array and smoothing a 3D array by one axis).

File: kaska/smoothn.py (hunk 2 above)
Comment (jgomezdans) [reply kaska/smoothn.py]: Those appear in the original paper. They're constants that have been pre-calculated. To be honest, I don't see anything wrong with having some of those numbers around. I know that they're often used in robust statistics, so I don't really see any problem with them. Basically, if you haven't read the paper, you shouldn't be touching the code.

File: kaska/smoothn.py [hunk 8]
```diff
@@ -523,8 +606,8 @@ def bound_calc(h_limit, rank):
 
 
 def initial_z(y, z0, is_weighted):
-    """
-    An initial conditions for z.
+    """An initial conditions for z.
```
Comment (ilectra) [inline kaska/smoothn.py]: Missed (the arguments/returns description in) one docstring.... :)

File: kaska/smoothn.py (hunk 2 above)
Comment (ilectra) [reply kaska/smoothn.py]: That's fine, I just meant putting the references next to the numbers, like Tim did in his last commit.

**Pull request merged**